### ISSUE TITLE
Objective-C: handling namespaces using objective-c++

### ISSUE THREAD

**Issue by dcow** [grpc/grpc#14036]
In our setup, we use folders to namespace our protos. Most of the tooling supports making sure paths between an include dir and the final source file are preserved in the generated output. The generated code is also namespaced using the relevant package option.

Enter objective-c, a language without namespaces or inner classes (similar issues exist if one were to venture into to generating c code). We have already encountered issues where two identical symbols are generated under objective-c. An objective-c class prefix option is available, and is what we use today, but I wonder if there's a more elegant way to handle codegen for objective-c. 

For example, objective-c and objective-c++ glue together pretty well. It might be interesting to have an option to codegen using objective-c++ so you can leverage namespaces. Most callers of the code would not even know they're calling objective-c++ instead of objective-c, and this is even how apple implements many of its own frameworks.

**Comment by dcow** [comment]
I should say I have not explored what it looks like simply calling the generated c++ code from objective-c. In my experience, I'm guessing you'd still need a thin wrapper layer to make sure all the c++-isms are hidden from callers lest their project explode into a full blown objective-c++ project (which we don't want).

**Comment by muxi** [comment]
I don't think it would work as you suggest. You cannot put an ObjC class in a C++ namespace in ObjC++ (see [this Wikipedia page](https://en.wikipedia.org/wiki/Objective-C#Language_variants) where it lists a bunch of restrictions).

AFAIK ObjC++ is more of an extension to ObjC compiler to recognize C++ syntax, rather than introducing new syntax that makes ObjC classes works with C++ features.

**Comment by dcow** [comment]
There is some new syntax from my experience, e.g. the ability to define class local variables to store private implementations without declaring the type in the class header (pimple idiom) etc. in an init block (which helps with hiding c++ from callers). But you are right I didn't know about the namespace restriction. However, that can be an implementation detail as far as this is concerned.

The idea would be to gen the actual classes using c++ (already done) and then also gen an objective-c++ wrapper that makes it easier to call into the objective-c++ classes while avoiding the namespace issues by bubbling namespace handling/resolution out to the wrapper. I opened this because I don't look forward to doing this by hand if I'm going to need to do it to generally resolve symbol conflicts, and of course I would be happy to help work this in if it's useful for anyone. 

The other issue that people will encounter if they were to venture into objective-c++ is that the c++ code will infect their project if it leaks through headers. That's the other role of objective-c++. You can call objective-c++ interfaces from objective-c and use the objective c++ layer to hide c++ semantics in .mm files while .h[h] files are purely objective-c. I haven't used the c++ codegen, but I have to imagine a intermediary layer would be useful as codegen because it's what most people would have to write themselves anyway.

Maybe I'll revisit this in the future with some actual examples if I end up trying to go this route.

**Comment by muxi** [comment]
I see your point here. Using C++ codegen and gRPC C++ library on iOS is one thing that we are working. We do not have any plan on writing ObjC wrapper on top of that right now, but once that is finished and you think this feature will be helpful contribution is always welcome.
